### PR TITLE
Feature/join room full

### PR DESCRIPTION
**Pull request by mikaeloduh**
## Why need this change? / Root cause: 
- Exception when room is full
## Changes made:
- Add maximum player validation 
## Test Scope / Change impact:
- `JoinRoomUsecase`
## Ticket
- resolve https://github.com/Game-as-a-Service/Lobby-Platform-Service/issues/54

### CODE REVIEW

**Comment by mikaeloduh** [comment]
第一次寫 Spring Kotlin，請小力鞭。

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 1]
```diff
@@ -35,6 +36,11 @@ class JoinRoomUsecase(
         }
     }
 
+    private fun Room.validateRoomIsFull() {
+        if (isFull())
+            throw PlatformException("The room (${roomId}) is full. Please select another room or try again later.")
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:41]: 大括號包起來

**Review by frankvicky — CHANGES_REQUESTED** [review]

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 2]
```diff
@@ -21,6 +21,7 @@ class JoinRoomUsecase(
         with(request) {
             val room = findRoomById(Room.Id(roomId))
             room.validateRoomPassword(password)
+            room.validateRoomIsFull()
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:24]: **validateFullRoom**

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt [hunk 3]
```diff
@@ -131,6 +132,31 @@ class RoomControllerTest @Autowired constructor(
             .thenShouldHaveRooms(request)
     }
 
+    @Test
+    fun givenRoomIsNotFull_whenUserTriesToJoin_ThenShouldSucceed() {
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt:136]: **givenRoomIsNotFull_whenUserJoinRoom_ThenShouldSucceed**

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt [hunk 4]
```diff
@@ -131,6 +132,31 @@ class RoomControllerTest @Autowired constructor(
...
+        val userA = testUser
+        val userB = createUser("2", "test2@mail.com", "1st_join_user")
+        givenTheHostCreatePublicRoom(userA)
+            .whenUserJoinTheRoom(userB)
+            .thenJoinRoomSuccessfully()
+    }
+
+    @Test
+    fun givenRoomIsFull_whenUserTriesToJoin_ThenShouldFail() {
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt:145]: **givenRoomIsFull_whenUserJoinRoom_ThenShouldFail**

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: domain/src/main/kotlin/tw/waterballsa/gaas/domain/Room.kt [hunk 5]
```diff
@@ -24,6 +24,10 @@ class Room(
         return this.password.equals(password)
     }
 
+    fun isFull(): Boolean {
+        return players.size >= maxPlayers
+    }
```
[the remark this answers is not in the document]
Comment (mikaeloduh) [reply domain/src/main/kotlin/tw/waterballsa/gaas/domain/Room.kt]: 修正在 1161adde82c1ae55cfed01a0b9b3b788fdbfab9f

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt (hunk 1 above)
Comment (mikaeloduh) [reply application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt]: 修正在 29abe34d4d5214158facb3df10a57e4db5fea6f4

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt (hunk 2 above)
Comment (mikaeloduh) [reply application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt]: 修正在 647f2d40f893e0873159eeb8c1125a2215012496

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt (hunk 4 above)
Comment (mikaeloduh) [reply spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt]: 修正在 1d0796e2f93188ae9ae6dac5969712d7cb751eac

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt (hunk 3 above)
Comment (mikaeloduh) [reply spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt]: 修正在 1d0796e2f93188ae9ae6dac5969712d7cb751eac

**Review by frankvicky — APPROVED** [review]

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Pull request merged**